src/main.rs

```diff
@@ -1,3 +1,829 @@
+
+#[derive(clap::Args)]
+struct CertifyArgs {
+    krate: String,
+    version: String,
+    message: String,
+}
+
+#[derive(clap::Args)]
+struct ForbidArgs {
+    krate: String,
+    version: String,
+    message: String,
+}
+
+#[derive(clap::Args)]
+struct SuggestArgs {}
+
+#[derive(clap::Args)]
+struct AuditsArgs {}
+
+#[derive(clap::Args)]
+struct HelpMarkdownArgs {}
+
+/// Logging verbosity levels
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, ArgEnum)]
+enum Verbose {
+    Off,
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+#[derive(serde::Deserialize)]
+struct Config {
+    // Reserved for future use, if not present version=1 assumed.
+    // (not sure whether this versions the format, or semantics, or...
+    // for now assuming this species global semantics of some kind.
+    version: Option<u64>,
+    store: Option<Store>,
+    audits: Option<Vec<String>>,
+}
+#[derive(serde::Deserialize)]
+struct Store {
+    path: Option<PathBuf>,
+}
+
+// FIXME: Probably want this to be a tree, and for queries to
+// be keyed off the "current package" but unclear what that would *mean*
+// so for now assume there is a linear chain of overrides. Possible
+// that any "tree-like" situation wants to be an error...
+
+/// All available configuration files, overlaying eachother.
+/// Generally contains: `[Default, Workspace, Package]`
+struct Configs(Vec<Config>);
+
+impl Configs {
+    fn store_path(&self) -> &Path {
+        // Last config gets priority to set this
+        for config in self.0.iter().rev() {
+            if let Some(store) = &config.store {
+                if let Some(path) = &store.path {
+                    return path;
+                }
+            }
+        }
+        unreachable!("Default config didn't define store.path???");
+    }
+    fn audits(&self) -> Vec<&str> {
+        self.0
+            .iter()
+            .flat_map(|cfg| cfg.audits.iter().flatten())
+            .map(|a| &**a)
+            .collect::<Vec<_>>()
+    }
+    fn version(&self) -> u64 {
+        // Last config gets priority to set this
+        for config in self.0.iter().rev() {
+            if let Some(ver) = config.version {
+                return ver;
+            }
+        }
+        unreachable!("Default config didn't define version???");
+    }
+}
+
+type Dependencies = BTreeMap<String, Vec<DependencyEntry>>;
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct Audited(Dependencies);
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct Unaudited(Dependencies);
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct Trusted(Dependencies);
+
+/// This is just a big vague ball initially. It's up to the Audits/Unuadited/Trusted wrappers
+/// to validate if it "makes sense" for their particular function.
+#[derive(serde::Serialize, serde::Deserialize)]
+struct DependencyEntry {
+    forbidden: Option<Version>,
+    version: Option<Version>,
+    delta: Option<Delta>,
+    who: Option<String>,
+}
```

Comment:
This is kinda a hack. Basically you are defining entries as implicitly enums, and the `toml` crate/format doesn't like that. To cope with it, I just have this function represent the union of all the enum variants and then do a post-process validation step (unimplemented, but conceptually doable).

Also since the different files seemed to basically be different subsets of the syntax and I'm already doing post-processing, it seemed reasonable to cram them all together.

The one problem with this definition is the one random place where you introduce `x.*` deps for just one of the files, which might necessitate making some of these into VersionReqs that I also post-process. (These are types from the `semver` crate since it's essentially the canonical source of truth on cargo version syntax, ideally whatever we do should interop with that.)

Reply:
My lean would be to have |struct AuditEntry| and |struct UnauditedEntry| which just map directly to the tables in the toml. We can define an enum over the two of them in situations where we need to abstract across them.